src/protocols/der/der.h

```diff
@@ -0,0 +1,171 @@
+	[FR_TYPE_DATE] = (bool []){[FR_DER_TAG_UTC_TIME] = true, [FR_DER_TAG_GENERALIZED_TIME] = true, [FR_DER_TAG_MAX] = false},
+	[FR_TYPE_TLV] = (bool []){[FR_DER_TAG_SEQUENCE] = true, [FR_DER_TAG_SET] = true, [FR_DER_TAG_MAX] = false},
+	[FR_TYPE_STRUCT] = (bool []){[FR_DER_TAG_BITSTRING] = true, [FR_DER_TAG_SEQUENCE] = true, [FR_DER_TAG_SET] = true, [FR_DER_TAG_MAX] = false},
+	[FR_TYPE_GROUP] = (bool []){[FR_DER_TAG_SEQUENCE] = true, [FR_DER_TAG_SET] = true, [FR_DER_TAG_MAX] = false}
+};
+
+/*
+ *	Return true if the given type can be encoded as the given tag.
+ * 		@param[in] type The fr_type to check.
+ * 		@param[in] tag The der tag to check.
+ * 		@return true if the type can be encoded as the given tag.
+ */
+static inline CC_HINT(always_inline) bool fr_type_to_der_tag_valid(fr_type_t type, fr_der_tag_num_t tag)
+{
+	return fr_type_to_der_tags[type][tag];
+}
+
+static int fr_type_to_der_tag_defaults[] = {
+	[FR_TYPE_NULL] = FR_DER_TAG_NULL,
+	[FR_TYPE_BOOL] = FR_DER_TAG_BOOLEAN,
+	[FR_TYPE_UINT8] = FR_DER_TAG_INTEGER,
+	[FR_TYPE_UINT16] = FR_DER_TAG_INTEGER,
+	[FR_TYPE_UINT32] = FR_DER_TAG_INTEGER,
+	[FR_TYPE_UINT64] = FR_DER_TAG_INTEGER,
+	[FR_TYPE_INT8] = FR_DER_TAG_INTEGER,
+	[FR_TYPE_INT16] = FR_DER_TAG_INTEGER,
+	[FR_TYPE_INT32] = FR_DER_TAG_INTEGER,
+	[FR_TYPE_INT64] = FR_DER_TAG_INTEGER,
+	[FR_TYPE_OCTETS] = FR_DER_TAG_OCTETSTRING,
+	[FR_TYPE_STRING] = FR_DER_TAG_UTF8_STRING,
+	[FR_TYPE_DATE] = FR_DER_TAG_GENERALIZED_TIME,
+	[FR_TYPE_TLV] = FR_DER_TAG_SEQUENCE,
+	[FR_TYPE_STRUCT] = FR_DER_TAG_SEQUENCE,
+	[FR_TYPE_GROUP] = FR_DER_TAG_SEQUENCE
+};
+
+static inline CC_HINT(always_inline) fr_der_tag_num_t fr_type_to_der_tag_default(fr_type_t type)
+{
+	return fr_type_to_der_tag_defaults[type];
+}
+
+#define DER_MAX_STR 16384
+
+#define DER_UTC_TIME_LEN 13	 //!< Length of the UTC time string.
+#define DER_GENERALIZED_TIME_LEN_MIN 15	 //!< Minimum length of the generalized time string.
+#define DER_GENERALIZED_TIME_PRECISION_MAX 4 //!< Maximum precision of the generalized time string.
+
+#define DER_TAG_CLASS_MASK 0xc0	 //!< Mask to extract the class from the tag.
+#define DER_TAG_CONSTRUCTED_MASK 0x20	 //!< Mask to check if the tag is constructed.
+#define DER_TAG_NUM_MASK 0x1f	 //!< Mask to extract the tag number from the tag.
+
+#define DER_MAX_TAG_NUM 0xfe * 8	 //!< Maximum tag number that can be encoded in a single byte.
+
+#define DER_TAG_CONTINUATION 0x1f	 //!< Mask to check if the tag is a continuation.
+
+#define DER_LEN_MULTI_BYTE 0x80	 //!< Mask to check if the length is multi-byte.
+
+#define DER_BOOLEAN_FALSE 0x00	 //!< DER encoded boolean false value.
+#define DER_BOOLEAN_TRUE 0xff	 //!< DER encoded boolean true value.
+
+typedef struct {
+	uint8_t tagnum;
```

Comment:
maybe format these so that the names are further to the right, and then all lined up vertically.  that makes it easier to read.